File: TopKFrequentElements.java
/**
 * LeetCode 347 Top K Frequent Elements
 */
class Solution {
    public List<Integer> topKFrequent(int[] nums, int k) {
        Map<Integer,Integer> map = new HashMap<>();
        for (int i = 0; i < nums.length; i++) {
            if (map.containsKey(nums[i])) {
                map.put(nums[i],map.get(nums[i]) + 1);
            } else {
                map.put(nums[i],1);
            }
        }

        PriorityQueue<Integer> priorityQueue = new PriorityQueue<>(new Comparator<Integer>() {
            @Override
            public int compare(Integer a, Integer b) {
                return map.get(a) - map.get(b);
            }
        });

        for (Integer key : map.keySet()) {
            if (priorityQueue.size() < k) {
                priorityQueue.add(key);
            } else if (map.get(key) > map.get(priorityQueue.peek())) {
                priorityQueue.remove();
                priorityQueue.add(key);
            }
        }

        List<Integer> res = new ArrayList<>();
        while (!priorityQueue.isEmpty()) {
            res.add(priorityQueue.remove());
        }

        return res;
    }
}

File: TopKFrequentElements.go
func topKFrequent(nums []int, k int) []int {
	container := make(map[int]int)
	for _,c := range nums {
		container[c]++
	}

	buckets := make([][]int,len(nums) + 1)
	for v,c := range container {
		if (len(buckets[c]) <= 0) {
			buckets[c] = make([]int,0)
		}
		buckets[c] = append(buckets[c],v)
	}

	res := make([]int,0)
	for i := len(buckets) - 1; i >= 0;i-- {
		if (len(buckets[i]) > 0) {
			res = append(res,buckets[i]...)
			if (len(res) >= k) {
				return res[0:k]
			}
		}
	}
	return res
}
